// @dynamic-labs/bitcoin/src/bitcoinWalletStandardWallets.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var app = require('@wallet-standard/app');

class BitcoinWalletStandardWallets {
    registerWallets() {
        this.wallets = app.getWallets();
    }
}
class BitcoinWalletStandardWalletsSingleton {
    constructor() {
        this.bitcoinWalletStandardWallets = new BitcoinWalletStandardWallets();
    }
    static get() {
        var _a;
        if (!((_a = BitcoinWalletStandardWalletsSingleton.instance) === null || _a === void 0 ? void 0 : _a.bitcoinWalletStandardWallets)) {
            BitcoinWalletStandardWalletsSingleton.instance =
                new BitcoinWalletStandardWalletsSingleton();
            BitcoinWalletStandardWalletsSingleton.instance.bitcoinWalletStandardWallets.registerWallets();
        }
        return BitcoinWalletStandardWalletsSingleton.instance
            .bitcoinWalletStandardWallets;
    }
    static getWallets() {
        var _a;
        return (_a = BitcoinWalletStandardWalletsSingleton.get().wallets) === null || _a === void 0 ? void 0 : _a.get();
    }
}

exports.BitcoinWalletStandardWallets = BitcoinWalletStandardWallets;
exports.BitcoinWalletStandardWalletsSingleton = BitcoinWalletStandardWalletsSingleton;
