// @dynamic-labs/bitcoin/src/index.js
'use client'
export { BitcoinWalletConnector } from './connectors/BitcoinWalletConnector.js';
import '../_virtual/_tslib.js';
import 'sats-connect';
import 'bitcoinjs-lib';
import '@dynamic-labs/wallet-connector-core';
import '@dynamic-labs/utils';
import '@dynamic-labs/wallet-book';
import { fetchBtcKitConnectors } from './utils/fetchBtcKitConnectors/fetchBtcKitConnectors.js';
export { fetchBtcKitConnectors } from './utils/fetchBtcKitConnectors/fetchBtcKitConnectors.js';
import { fetchSatsConnectConnectors } from './utils/fetchSatsConnectConnectors/fetchSatsConnectConnectors.js';
export { fetchSatsConnectConnectors } from './utils/fetchSatsConnectConnectors/fetchSatsConnectConnectors.js';
export { hasSatsConnectFeature } from './utils/hasSatsConnectFeature.js';
export { supportsSatsConnect } from './utils/supportsSatsConnect.js';
export { getMempoolApiUrl } from './utils/getMempoolApiUrl.js';
import { PhantomConnector } from './connectors/PhantomConnector/PhantomConnector.js';
import { OkxConnector } from './connectors/OkxConnector/OkxConnector.js';
import { UnisatConnector } from './connectors/UnisatConnector/UnisatConnector.js';
import { UnknownInjectedConnector } from './connectors/UnknownInjected/UnknownInjected.js';
import { FallbackBitcoinConnector } from './connectors/FallbackBitcoinConnector/FallbackBitcoinConnector.js';

const BitcoinWalletConnectors = (props) => [
    ...fetchSatsConnectConnectors(props),
    ...fetchBtcKitConnectors(props),
    PhantomConnector,
    OkxConnector,
    UnisatConnector,
    UnknownInjectedConnector,
    FallbackBitcoinConnector,
];

export { BitcoinWalletConnectors };
