// @dynamic-labs/ethereum/src/index.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('./polyfills.cjs');
var turnkey = require('@dynamic-labs/turnkey');
var PhantomEvm = require('./injected/PhantomEvm.cjs');
var ExodusEvm = require('./injected/ExodusEvm.cjs');
var UnknownInjected = require('./injected/UnknownInjected.cjs');
var FallbackEvmConnector = require('./injected/FallbackEvmConnector.cjs');
var fetchInjectedWalletConnectors = require('./injected/fetchInjectedWalletConnectors.cjs');
require('./walletConnect/walletConnect.cjs');
var fetchWalletConnectWallets = require('./walletConnect/fetchWalletConnectWallets.cjs');
var coinbase = require('./coinbase/coinbase.cjs');
require('@dynamic-labs/rpc-provider-ethereum');
var EthWalletConnector = require('./EthWalletConnector.cjs');
var ethProviderHelper = require('./ethProviderHelper.cjs');
var constants = require('./constants.cjs');
var isEthWalletConnector = require('./utils/isEthWalletConnector/isEthWalletConnector.cjs');
var getNameservice = require('./utils/getNameservice.cjs');

const EthereumWalletConnectors = (props) => [
    ...fetchInjectedWalletConnectors.injectedWalletOverrides,
    ...fetchInjectedWalletConnectors.fetchInjectedWalletConnector(props),
    ...fetchWalletConnectWallets.fetchWalletConnectWallets(props),
    ...turnkey.TurnkeyEVMWalletConnectors(props),
    coinbase.Coinbase,
    UnknownInjected.UnknownInjected,
    FallbackEvmConnector.FallbackEvmConnector,
    fetchWalletConnectWallets.getWalletConnectConnector(),
];

exports.PhantomEvm = PhantomEvm.PhantomEvm;
exports.ExodusEvm = ExodusEvm.ExodusEvm;
exports.UnknownInjected = UnknownInjected.UnknownInjected;
exports.FallbackEvmConnector = FallbackEvmConnector.FallbackEvmConnector;
exports.fetchInjectedWalletConnector = fetchInjectedWalletConnectors.fetchInjectedWalletConnector;
exports.injectedWalletOverrides = fetchInjectedWalletConnectors.injectedWalletOverrides;
exports.EthWalletConnector = EthWalletConnector.EthWalletConnector;
exports.EthProviderHelper = ethProviderHelper.EthProviderHelper;
exports.INFURA_ID = constants.INFURA_ID;
exports.isEthWalletConnector = isEthWalletConnector.isEthWalletConnector;
exports.getNameservice = getNameservice.getNameservice;
exports.EthereumWalletConnectors = EthereumWalletConnectors;
