// @dynamic-labs/solana/src/injected/InjectedWalletBase.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../_virtual/_tslib.cjs');
var walletBook = require('@dynamic-labs/wallet-book');
var utils = require('@dynamic-labs/utils');
var solProviderHelper = require('../solProviderHelper.cjs');
var solWalletConnector = require('../solWalletConnector.cjs');
var SignMessageNotSupportedError = require('../errors/SignMessageNotSupportedError.cjs');

class InjectedWalletBase extends solWalletConnector.SolWalletConnector {
    getSolProviderHelper() {
        if (!this.wallet) {
            this.wallet = walletBook.findWalletBookWallet(this.walletBook, this.key);
        }
        if (this.wallet && !this.solProviderHelper) {
            this.solProviderHelper = new solProviderHelper.SolProviderHelper(this.wallet);
        }
        return this.solProviderHelper;
    }
    getProvider() {
        var _a;
        return (_a = this.getSolProviderHelper()) === null || _a === void 0 ? void 0 : _a.getInstalledProvider();
    }
    setupEventListeners() {
        var _a;
        (_a = this.getSolProviderHelper()) === null || _a === void 0 ? void 0 : _a._setupEventListeners(this);
    }
    teardownEventListeners() {
        var _a;
        (_a = this.getSolProviderHelper()) === null || _a === void 0 ? void 0 : _a._teardownEventListeners();
    }
    connect() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            var _a;
            yield ((_a = this.getSolProviderHelper()) === null || _a === void 0 ? void 0 : _a.connect());
        });
    }
    getSigner() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            var _a;
            return (_a = this.getSolProviderHelper()) === null || _a === void 0 ? void 0 : _a.connect();
        });
    }
    isInstalledOnBrowser() {
        var _a;
        return Boolean((_a = this.getSolProviderHelper()) === null || _a === void 0 ? void 0 : _a.isInstalledHelper());
    }
    getAddress() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            if (!this.isInstalledOnBrowser()) {
                if (utils.isMobile() &&
                    ((_a = this.wallet) === null || _a === void 0 ? void 0 : _a.mobile) &&
                    this.wallet.mobile.inAppBrowser) {
                    const inAppBrowserCompiledTemplate = utils.template(this.wallet.mobile.inAppBrowser);
                    const deepLink = inAppBrowserCompiledTemplate({
                        encodedDappURI: encodeURIComponent(window.location.toString()),
                    });
                    window.location.assign(deepLink);
                }
                return;
            }
            return (_b = this.getSolProviderHelper()) === null || _b === void 0 ? void 0 : _b.getAddress();
        });
    }
    signMessage(messageToSign) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const walletAddress = yield ((_a = this.getSolProviderHelper()) === null || _a === void 0 ? void 0 : _a.getAddress());
            if (walletAddress && this.isLedgerAddress(walletAddress)) {
                throw new SignMessageNotSupportedError.SignMessageNotSupportedError(this.name);
            }
            return (_b = this.getSolProviderHelper()) === null || _b === void 0 ? void 0 : _b.signMessage(messageToSign);
        });
    }
    getConnectedAccounts() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            return (_b = (_a = this.getSolProviderHelper()) === null || _a === void 0 ? void 0 : _a.getConnectedAccounts()) !== null && _b !== void 0 ? _b : [];
        });
    }
}

exports.InjectedWalletBase = InjectedWalletBase;
