// @dynamic-labs/solana/src/injected/PhantomInjected.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../_virtual/_tslib.cjs');
var utils = require('@dynamic-labs/utils');
var walletBook = require('@dynamic-labs/wallet-book');
var InjectedWalletBase = require('./InjectedWalletBase.cjs');

class PhantomInjected extends InjectedWalletBase.InjectedWalletBase {
    constructor(opts) {
        super(opts);
        this.name = 'Phantom';
        this.overrideKey = 'phantom';
        this.wallet = walletBook.findWalletBookWallet(this.walletBook, this.key);
    }
    getAddress() {
        const _super = Object.create(null, {
            getAddress: { get: () => super.getAddress }
        });
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            if (this.isInstalledOnBrowser()) {
                return _super.getAddress.call(this);
            }
            if (utils.isMobile()) {
                utils.handleMobileWalletRedirect({
                    nativeLink: 'phantom://browse',
                    universalLink: 'https://phantom.app/ul/browse',
                });
            }
            return undefined;
        });
    }
}

exports.PhantomInjected = PhantomInjected;
