// @dynamic-labs/solana/src/Phantom.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../_virtual/_tslib.cjs');
var walletBook = require('@dynamic-labs/wallet-book');
var utils = require('@dynamic-labs/utils');
var solWalletConnector = require('./solWalletConnector.cjs');
var PhantomRedirect = require('./phantomRedirect/PhantomRedirect.cjs');
var PhantomInjected = require('./injected/PhantomInjected.cjs');

class Phantom extends solWalletConnector.SolWalletConnector {
    constructor(opts) {
        super(opts);
        this.name = 'Phantom';
        this.overrideKey = 'phantom';
        this.wallet = walletBook.findWalletBookWallet(this.walletBook, this.key);
    }
    connect() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            yield this.getMobileOrInstalledWallet().connect();
        });
    }
    getSigner() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            return this.getMobileOrInstalledWallet().getSigner();
        });
    }
    getMobileOrInstalledWallet() {
        const phantomInjected = new PhantomInjected.PhantomInjected(this.constructorProps);
        if (!utils.isMobile() || phantomInjected.isInstalledOnBrowser()) {
            return phantomInjected;
        }
        if (this.constructorProps.mobileExperience === 'redirect') {
            return new PhantomRedirect.PhantomRedirect(this.constructorProps);
        }
        else {
            return new PhantomInjected.PhantomInjected(this.constructorProps);
        }
    }
}

exports.Phantom = Phantom;
