// @dynamic-labs/bitcoin/src/connectors/FallbackBitcoinConnector/FallbackBitcoinConnector.js
'use client'
import { __awaiter } from '../../../_virtual/_tslib.js';
import { BitcoinWalletConnector } from '../BitcoinWalletConnector.js';

class FallbackBitcoinConnector extends BitcoinWalletConnector {
    constructor(opts) {
        super(Object.assign(Object.assign({}, opts), { overrideKey: 'fallbackconnector' }));
        this.name = 'Fallback Connector';
        this.overrideKey = 'fallbackconnector';
        this.isAvailable = false;
        this.canFetchConnectedAccounts = false;
    }
    getAddress() {
        return __awaiter(this, void 0, void 0, function* () {
            return;
        });
    }
    signPsbt(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _request) {
        return __awaiter(this, void 0, void 0, function* () {
            return;
        });
    }
    sendBitcoin(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _transaction) {
        return __awaiter(this, void 0, void 0, function* () {
            return;
        });
    }
}

export { FallbackBitcoinConnector };
